[WTF] Modernize wtf/MathExtras.h

weinig opened the pull request:
#### 38b34df1ab771b304387f2c75c6f1a775f22028f
<pre>
[WTF] Modernize wtf/MathExtras.h
<a href="https://bugs.webkit.org/show_bug.cgi?id=296832">https://bugs.webkit.org/show_bug.cgi?id=296832</a>

Reviewed by Darin Adler.

Does a pass of modernizing wtf/MathExtras.h by utilizing
concepts/requires and adopting constexpr more broadly.

* Source/WTF/wtf/MathExtras.h:
(WTF::isNaNConstExpr):
(WTF::fabsConstExpr):
(defaultMinimumForClamp):
(defaultMaximumForClamp):
(clampTo):
(clampToInteger):
(clampToUnsigned):
(clampToFloat):
(clampToPositiveInteger):
(isWithinIntRange):
(normalizedFloat):
(divideRoundedUp):
(timesThreePlusOneDividedByTwo):
(isNotZeroAndOrdered):
(isZeroOrUnordered):
(isGreaterThanNonZeroPowerOfTwo):
(WTF::fastLog2):
(WTF::safeFPDivision):
(WTF::areEssentiallyEqual):
(WTF::nanPropagatingMin):
(WTF::nanPropagatingMax):
(WTF::isIntegral):
(WTF::incrementWithSaturation):
(WTF::leftShiftWithSaturation):
(WTF::nonEmptyRangesOverlap):
(WTF::rangesOverlap):
(WTF::shuffleVector):
(WTF::isIdentical):
(WTF::isRepresentableAsImpl):
(WTF::isRepresentableAs):
(WTF::roundUpToMultipleOf):
(WTF::roundUpToMultipleOfNonPowerOfTwo):
(WTF::distanceToMultipleOf):
(WTF::roundDownToMultipleOf):
(sizeof): Deleted.

Canonical link: <a href="https://commits.webkit.org/298169@main">https://commits.webkit.org/298169@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8adb9e52adbbff4cd9333531e11cd9ad1e7102cc

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/114492 "35 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/34238 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/24701 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/120658 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/65212 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/34867 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/42798 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/87016 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/41921 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/117440 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/27780 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/102821 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/67407 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/26964 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/20948 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/64338 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/106888 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/97157 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/21062 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/123863 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/113055 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/41505 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/30973 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/95836 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/41882 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/99012 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/95620 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24363 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/40781 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/18618 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/37540 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/41384 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/46894 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/137272 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/40972 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/36707 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/44279 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/42722 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WTF/wtf/MathExtras.h:
```diff
@@ -238,36 +263,34 @@ clampTo(SourceType value)
 {
-    static_assert(std::numeric_limits<T>::is_integer, "T must be an integer.");
-
     const T intMax = static_cast<unsigned>(std::numeric_limits<int>::max());
```
```suggestion
    constexpr T intMax = static_cast<unsigned>(std::numeric_limits<int>::max());
```

darinadler commented on Source/WTF/wtf/MathExtras.h:
```diff
@@ -472,33 +495,33 @@ inline typename std::enable_if<std::is_floating_point<T>::value, bool>::type are
+constexpr void incrementWithSaturation(T& value)
 {
     if (value != std::numeric_limits<T>::max())
         value++;
```
```suggestion
        ++value;
```

darinadler commented on Source/WTF/wtf/MathExtras.h:
```diff
@@ -657,72 +680,43 @@ inline uint32_t reverseBits32(uint32_t value)
 {
     BitsType leftBits = std::bit_cast<BitsType>(left);
     BitsType rightBits = std::bit_cast<BitsType>(right);
     return leftBits == rightBits;
```
```suggestion
    return std::bit_cast<BitsType>(left) == std::bit_cast<BitsType>(right);
```

darinadler commented on Source/WTF/wtf/MathExtras.h:
```diff
@@ -779,14 +773,14 @@ template<size_t divisor> constexpr size_t roundUpToMultipleOf(size_t x)
 }
 
-template<size_t divisor, typename T> inline constexpr T* roundUpToMultipleOf(T* x)
+template<size_t divisor, typename T> constexpr T* roundUpToMultipleOf(T* x)
```
```suggestion
template<size_t divisor, typename T>
constexpr T* roundUpToMultipleOf(T* x)
```

weinig committed:
[WTF] Modernize wtf/MathExtras.h
https://bugs.webkit.org/show_bug.cgi?id=296832

Reviewed by Darin Adler.

Does a pass of modernizing wtf/MathExtras.h by utilizing
concepts/requires and adopting constexpr more broadly.

* Source/WTF/wtf/MathExtras.h:
(WTF::isNaNConstExpr):
(WTF::fabsConstExpr):
(defaultMinimumForClamp):
(defaultMaximumForClamp):
(clampTo):
(clampToInteger):
(clampToUnsigned):
(clampToFloat):
(clampToPositiveInteger):
(isWithinIntRange):
(normalizedFloat):
(divideRoundedUp):
(timesThreePlusOneDividedByTwo):
(isNotZeroAndOrdered):
(isZeroOrUnordered):
(isGreaterThanNonZeroPowerOfTwo):
(WTF::fastLog2):
(WTF::safeFPDivision):
(WTF::areEssentiallyEqual):
(WTF::nanPropagatingMin):
(WTF::nanPropagatingMax):
(WTF::isIntegral):
(WTF::incrementWithSaturation):
(WTF::leftShiftWithSaturation):
(WTF::nonEmptyRangesOverlap):
(WTF::rangesOverlap):
(WTF::shuffleVector):
(WTF::isIdentical):
(WTF::isRepresentableAsImpl):
(WTF::isRepresentableAs):
(WTF::roundUpToMultipleOf):
(WTF::roundUpToMultipleOfNonPowerOfTwo):
(WTF::distanceToMultipleOf):
(WTF::roundDownToMultipleOf):
(sizeof): Deleted.

Canonical link: https://commits.webkit.org/298169@main
